File: ts/indicators/jean-philippe-poton/fractal-bands-hybride-adaptive/fractal-bands-hybride-adaptive.ts
import { buildMetadata } from '../../core/build-metadata';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { LineIndicator } from '../../core/line-indicator';
import { Band } from '../../core/outputs/band';
import { Scalar } from '../../../entities/scalar';
import { Bar } from '../../../entities/bar';
import { Quote } from '../../../entities/quote';
import { Trade } from '../../../entities/trade';
import { FractalBandsHybrideAdaptiveParams } from './params';

/** Function to calculate mnemonic of a __FractalBandsHybrideAdaptive__ indicator. */
export const fractalBandsHybrideAdaptiveMnemonic = (params: FractalBandsHybrideAdaptiveParams): string =>
    'fbanha('.concat(
        params.period.toString(), ',',
        params.normalSpeedFallback.toString(), ',',
        params.alpha.toString(), ',',
        params.nyquist.toString(), ',',
        params.alphaHP.toString(),
        componentTripleMnemonic(params.barComponent, params.quoteComponent, params.tradeComponent),
        ')');

/** Fractal Bands Hybride Adaptive indicator. */
export class FractalBandsHybrideAdaptive extends LineIndicator {
    private window: Array<number>;
    private _closes: Array<number>;
    private _period: number;
    private _windowSize: number;
    private _normalSpeedFallback: number;
    private _alpha: number;
    private _nyquist: number;
    private _alphaHP: number;
    private windowCount: number;
    private logDenom: number;
    private ln2: number;
    private invPeriodSq: number;
    private _frasma2: number;
    private _upperBand: number;
    private _lowerBand: number;
    // Ehlers CyclePeriod buffers.
    private smoothBuf: Array<number>;
    private cycleBuf: Array<number>;
    private q1Buf: Array<number>;
    private i1Buf: Array<number>;
    private dpBuf: Array<number>;
    private instPeriodBuf: Array<number>;

    /**
     * Constructs an instance given parameters.
     **/
    public constructor(params: FractalBandsHybrideAdaptiveParams) {
        super();
        const period = Math.floor(params.period);
        const normalSpeedFallback = Math.floor(params.normalSpeedFallback);
        const alpha = params.alpha;
        const nyquist = params.nyquist;
        const alphaHP = params.alphaHP;

        if (period < 2) {
            throw new Error('period should be greater than 1');
        }

        if (normalSpeedFallback < 1) {
            throw new Error('normal_speed_fallback should be greater than 0');
        }

        if (alpha <= 0.0) {
            throw new Error('alpha should be greater than 0');
        }

        if (nyquist <= 0.0) {
            throw new Error('nyquist should be greater than 0');
        }

        if (alphaHP <= 0.0 || alphaHP >= 1.0) {
            throw new Error('alpha_hp should be between 0 and 1');
        }

        this.mnemonic = fractalBandsHybrideAdaptiveMnemonic(params);
        this.description = 'Fractal bands hybride adaptive ' + this.mnemonic;
        this.barComponent = params.barComponent;
        this.quoteComponent = params.quoteComponent;
        this.tradeComponent = params.tradeComponent;
        this.window = new Array<number>(period + 1);
        this._closes = [];
        this._period = period;
        this._windowSize = period + 1;
        this._normalSpeedFallback = normalSpeedFallback;
        this._alpha = alpha;
        this._nyquist = nyquist;
        this._alphaHP = alphaHP;
        this.windowCount = 0;
        this.primed = false;
        this.logDenom = Math.log(2.0 * (period - 1));
        this.ln2 = Math.log(2.0);
        this.invPeriodSq = 1.0 / (period * period);
        this._frasma2 = Number.NaN;
        this._upperBand = Number.NaN;
        this._lowerBand = Number.NaN;
        this.smoothBuf = [];
        this.cycleBuf = [];
        this.q1Buf = [];
        this.i1Buf = [];
        this.dpBuf = [];
        this.instPeriodBuf = [];
    }

    /** Describes the output data of the indicator. */
    public metadata(): IndicatorMetadata {
        return buildMetadata(
            IndicatorIdentifier.FractalBandsHybrideAdaptive,
            this.mnemonic,
            this.description,
            [
                { mnemonic: this.mnemonic, description: this.description },
                { mnemonic: this.mnemonic + ' upper', description: this.description + ' Upper Band' },
                { mnemonic: this.mnemonic + ' lower', description: this.description + ' Lower Band' },
                { mnemonic: this.mnemonic + ' band', description: this.description + ' Band' },
            ],
        );
    }

    /** Compute Ehlers CyclePeriod for the current bar. */
    private getCyclePeriod(): number {
        const t = this._closes.length - 1;
        const prices = this._closes;

        // Extend buffers to index t.
        while (this.smoothBuf.length <= t) { this.smoothBuf.push(0.0); }
        while (this.cycleBuf.length <= t) { this.cycleBuf.push(0.0); }
        while (this.q1Buf.length <= t) { this.q1Buf.push(0.0); }
        while (this.i1Buf.length <= t) { this.i1Buf.push(0.0); }
        while (this.dpBuf.length <= t) { this.dpBuf.push(0.0); }
        while (this.instPeriodBuf.length <= t) { this.instPeriodBuf.push(6.0); }

        if (t < 6) {
            return Number.NaN;
        }

        // 4-bar weighted smoother.
        this.smoothBuf[t] = (prices[t] + 2.0 * prices[t - 1] +
            2.0 * prices[t - 2] + prices[t - 3]) / 6.0;

        // High-pass filter.
        const alphaHP = this._alphaHP;
        const hpCoeff = (1.0 - 0.5 * alphaHP) * (1.0 - 0.5 * alphaHP);
        const oneMinusAlpha = 1.0 - alphaHP;

        this.cycleBuf[t] = hpCoeff * (this.smoothBuf[t] - 2.0 * this.smoothBuf[t - 1] + this.smoothBuf[t - 2]) +
            2.0 * oneMinusAlpha * this.cycleBuf[t - 1] - oneMinusAlpha * oneMinusAlpha * this.cycleBuf[t - 2];

        // Quadrature component.
        this.q1Buf[t] = (0.0962 * this.cycleBuf[t] + 0.5769 * this.cycleBuf[t - 2] -
            0.5769 * this.cycleBuf[t - 4] - 0.0962 * this.cycleBuf[t - 6]) *
            (0.5 + 0.08 * this.instPeriodBuf[t - 1]);

        // In-phase component.
        this.i1Buf[t] = this.cycleBuf[t - 3];

        // Smooth I and Q with EMA.
        if (t > 6) {
            this.i1Buf[t] = 0.15 * this.i1Buf[t] + 0.85 * this.i1Buf[t - 1];
            this.q1Buf[t] = 0.15 * this.q1Buf[t] + 0.85 * this.q1Buf[t - 1];
        }

        // Compute delta phase.
        let dp: number;
        if (Math.abs(this.i1Buf[t]) > 1e-10) {
            dp = Math.atan(this.q1Buf[t] / this.i1Buf[t]);
        } else {
            dp = this.dpBuf[t - 1];
        }

        // Clamp delta phase.
        if (dp < 0.1) { dp = 0.1; }
        if (dp > 1.1) { dp = 1.1; }
        this.dpBuf[t] = dp;

        // Median delta phase over 5 bars.
        let medianDP: number;
        if (t >= 10) {
            const w = [this.dpBuf[t - 4], this.dpBuf[t - 3], this.dpBuf[t - 2],
                this.dpBuf[t - 1], this.dpBuf[t]];
            w.sort((a, b) => a - b);
            medianDP = w[2];
        } else {
            medianDP = dp;
        }

        // Instantaneous period.
        let dc: number;
        if (Math.abs(medianDP) > 1e-10) {
            dc = 6.2832 / medianDP + 0.5;
        } else {
            dc = this.instPeriodBuf[t - 1];
        }

        // Clamp and smooth.
        if (dc < 6.0) { dc = 6.0; }
        if (dc > 50.0) { dc = 50.0; }
        this.instPeriodBuf[t] = 0.33 * dc + 0.67 * this.instPeriodBuf[t - 1];

        return this.instPeriodBuf[t];
    }

    /** Updates the indicator and returns the FRASMA2 value. */
    public update(sample: number): number {
        if (Number.isNaN(sample)) {
            return sample;
        }

        const period = this._period;
        const windowSize = this._windowSize;

        // Accumulate close history.
        this._closes.push(sample);

        // Update Ehlers CyclePeriod.
        const cp = this.getCyclePeriod();

        // Fill the FGDI window (period+1 elements).
        if (this.primed) {
            for (let i = 0; i < windowSize - 1; i++) {
                this.window[i] = this.window[i + 1];
            }
            this.window[windowSize - 1] = sample;
        } else {
            this.window[this.windowCount] = sample;
            this.windowCount++;

            if (this.windowCount < windowSize) {
                return Number.NaN;
            }

            this.primed = true;
        }

        // FGDI computation over period+1 points.
        let priceMax = this.window[0];
        let priceMin = this.window[0];

        for (let k = 1; k < windowSize; k++) {
            if (this.window[k] > priceMax) { priceMax = this.window[k]; }
            if (this.window[k] < priceMin) { priceMin = this.window[k]; }
        }

        const priceRange = priceMax - priceMin;
        let fgdi: number;

        if (priceRange < 1e-10) {
            fgdi = 1.0;
        } else {
            let length = 0.0;
            for (let i = 1; i < windowSize; i++) {
                const normCur = (this.window[i] - priceMin) / priceRange;
                const normPrev = (this.window[i - 1] - priceMin) / priceRange;
                const diff = normCur - normPrev;
                length += Math.sqrt(diff * diff + this.invPeriodSq);
            }
            fgdi = 1.0 + (Math.log(length) + this.ln2) / this.logDenom;
        }

        // Hurst exponent.
        let hurst = 2.0 - fgdi;
        if (hurst < 0.01) { hurst = 0.01; }

        const trailDim = 1.0 / hurst;
        const beta = trailDim / 2.0;

        // Adaptive normal_speed from CyclePeriod.
        let ns: number;
        if (Number.isNaN(cp) || cp < 1.0) {
            ns = this._normalSpeedFallback;
        } else {
            ns = cp * this._nyquist;
        }

        const speed = Math.max(Math.round(ns * beta), 1);

        // FRASMA2: SMA of close over 'speed' bars ending at current position.
        const nCloses = this._closes.length;
        if (speed > nCloses) {
            this._frasma2 = Number.NaN;
            this._upperBand = Number.NaN;
            this._lowerBand = Number.NaN;
            return Number.NaN;
        }

        let smaSum = 0.0;
        for (let k = nCloses - speed; k < nCloses; k++) {
            smaSum += this._closes[k];
        }
        const frasma2Val = smaSum / speed;

        // Deviation over the last period closes.
        let sqSum = 0.0;
        const devStart = Math.max(nCloses - period, 0);
        for (let k = devStart; k < nCloses; k++) {
            const res = this._closes[k] - frasma2Val;
            sqSum += res * res;
        }
        const deviation = 2.0 * Math.sqrt(sqSum / period);

        // Fractal bands.
        const bandMult = deviation * Math.pow(this._alpha, hurst);
        const upperBand = frasma2Val + bandMult;
        const lowerBand = frasma2Val - bandMult;

        this._frasma2 = frasma2Val;
        this._upperBand = upperBand;
        this._lowerBand = lowerBand;

        return frasma2Val;
    }

    /** Updates and returns all three outputs: [frasma2, upperBand, lowerBand]. */
    public updateAll(sample: number): [number, number, number] {
        const frasma2 = this.update(sample);
        return [frasma2, this._upperBand, this._lowerBand];
    }

    /** Returns the last computed FRASMA2 value. */
    public get frasma2Value(): number { return this._frasma2; }

    /** Returns the last computed upper band value. */
    public get upperBandValue(): number { return this._upperBand; }

    /** Returns the last computed lower band value. */
    public get lowerBandValue(): number { return this._lowerBand; }

    /** Updates the indicator given the next scalar sample. */
    public updateScalar(sample: Scalar): IndicatorOutput {
        const [frasma2, upper, lower] = this.updateAll(sample.value);

        const s0 = new Scalar(); s0.time = sample.time; s0.value = frasma2;
        const s1 = new Scalar(); s1.time = sample.time; s1.value = upper;
        const s2 = new Scalar(); s2.time = sample.time; s2.value = lower;

        const band = new Band();
        band.time = sample.time;
        if (isNaN(lower) || isNaN(upper)) {
            band.lower = NaN;
            band.upper = NaN;
        } else {
            band.lower = lower;
            band.upper = upper;
        }

        return [s0, s1, s2, band];
    }

    /** Updates the indicator given the next bar sample. */
    public updateBar(sample: Bar): IndicatorOutput {
        const v = this.barComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next quote sample. */
    public updateQuote(sample: Quote): IndicatorOutput {
        const v = this.quoteComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }

    /** Updates the indicator given the next trade sample. */
    public updateTrade(sample: Trade): IndicatorOutput {
        const v = this.tradeComponentFunc(sample);
        const scalar = new Scalar();
        scalar.time = sample.time;
        scalar.value = v;
        return this.updateScalar(scalar);
    }
}

File: go/indicators/jeanphilippepoton/fractalbandshybrideadaptive/fractalbandshybrideadaptive.go
package fractalbandshybrideadaptive

import (
	"fmt"
	"math"
	"sync"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/core/outputs"
)

// FractalBandsHybrideAdaptive computes the Fractal Bands Hybride Adaptive indicator.
//
// Hybrid variant of Fractal Bands that replaces fixed normal_speed with
// Ehlers' CyclePeriod indicator output multiplied by a Nyquist factor,
// making the FRASMA2 doubly adaptive to both fractal dimension and
// dominant market cycle.
//
// The indicator is not primed during the first `period` updates.
type FractalBandsHybrideAdaptive struct {
	mu sync.RWMutex
	core.LineIndicator
	barFunc            entities.BarFunc
	quoteFunc          entities.QuoteFunc
	tradeFunc          entities.TradeFunc
	window             []float64
	closes             []float64
	period             int
	windowSize         int
	normalSpeedFallback int
	alpha              float64
	nyquist            float64
	alphaHP            float64
	windowCount        int
	primed             bool
	logDenom           float64
	ln2                float64
	invPeriodSq        float64
	// Ehlers CyclePeriod buffers.
	smoothBuf    []float64
	cycleBuf     []float64
	q1Buf        []float64
	i1Buf        []float64
	dpBuf        []float64
	instPeriodBuf []float64
	// Last computed values.
	frasma2   float64
	upperBand float64
	lowerBand float64
}

// NewFractalBandsHybrideAdaptive returns an instance of the indicator created using supplied parameters.
func NewFractalBandsHybrideAdaptive(p *Params) (*FractalBandsHybrideAdaptive, error) {
	const (
		invalid = "invalid fractal bands hybride adaptive parameters"
		fmts    = "%s: %s"
		fmtw    = "%s: %w"
		fmtn    = "fbanha(%d,%d,%g,%g,%g%s)"
	)

	period := p.Period
	if period < 2 {
		return nil, fmt.Errorf(fmts, invalid, "period should be greater than 1")
	}

	normalSpeedFallback := p.NormalSpeedFallback
	if normalSpeedFallback < 1 {
		return nil, fmt.Errorf(fmts, invalid, "normal_speed_fallback should be greater than 0")
	}

	alpha := p.Alpha
	if alpha <= 0.0 {
		return nil, fmt.Errorf(fmts, invalid, "alpha should be greater than 0")
	}

	nyquist := p.Nyquist
	if nyquist <= 0.0 {
		return nil, fmt.Errorf(fmts, invalid, "nyquist should be greater than 0")
	}

	alphaHP := p.AlphaHP
	if alphaHP <= 0.0 || alphaHP >= 1.0 {
		return nil, fmt.Errorf(fmts, invalid, "alpha_hp should be between 0 and 1")
	}

	bc := p.BarComponent
	if bc == 0 {
		bc = entities.DefaultBarComponent
	}

	qc := p.QuoteComponent
	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	tc := p.TradeComponent
	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	var (
		err       error
		barFunc   entities.BarFunc
		quoteFunc entities.QuoteFunc
		tradeFunc entities.TradeFunc
	)

	if barFunc, err = entities.BarComponentFunc(bc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if quoteFunc, err = entities.QuoteComponentFunc(qc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	if tradeFunc, err = entities.TradeComponentFunc(tc); err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	mnemonic := fmt.Sprintf(fmtn, period, normalSpeedFallback, alpha, nyquist, alphaHP,
		core.ComponentTripleMnemonic(bc, qc, tc))
	desc := "Fractal bands hybride adaptive " + mnemonic

	windowSize := period + 1

	f := &FractalBandsHybrideAdaptive{
		barFunc:            barFunc,
		quoteFunc:          quoteFunc,
		tradeFunc:          tradeFunc,
		window:             make([]float64, windowSize),
		closes:             make([]float64, 0, 256),
		period:             period,
		windowSize:         windowSize,
		normalSpeedFallback: normalSpeedFallback,
		alpha:              alpha,
		nyquist:            nyquist,
		alphaHP:            alphaHP,
		logDenom:           math.Log(2.0 * float64(period-1)),
		ln2:               math.Log(2.0),
		invPeriodSq:        1.0 / float64(period*period),
		smoothBuf:          make([]float64, 0, 256),
		cycleBuf:           make([]float64, 0, 256),
		q1Buf:              make([]float64, 0, 256),
		i1Buf:              make([]float64, 0, 256),
		dpBuf:              make([]float64, 0, 256),
		instPeriodBuf:      make([]float64, 0, 256),
		frasma2:           math.NaN(),
		upperBand:         math.NaN(),
		lowerBand:         math.NaN(),
	}

	f.LineIndicator = core.NewLineIndicator(mnemonic, desc, barFunc, quoteFunc, tradeFunc, f.Update)

	return f, nil
}

// IsPrimed indicates whether the indicator is primed.
func (f *FractalBandsHybrideAdaptive) IsPrimed() bool {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.primed
}

// Metadata describes the output data of the indicator.
func (f *FractalBandsHybrideAdaptive) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.FractalBandsHybrideAdaptive,
		f.LineIndicator.Mnemonic,
		f.LineIndicator.Description,
		[]core.OutputText{
			{Mnemonic: f.LineIndicator.Mnemonic, Description: f.LineIndicator.Description},
			{Mnemonic: f.LineIndicator.Mnemonic + " upper", Description: f.LineIndicator.Description + " Upper Band"},
			{Mnemonic: f.LineIndicator.Mnemonic + " lower", Description: f.LineIndicator.Description + " Lower Band"},
			{Mnemonic: f.LineIndicator.Mnemonic + " band", Description: f.LineIndicator.Description + " Band"},
		},
	)
}

// getCyclePeriod computes the Ehlers CyclePeriod for the current bar.
func (f *FractalBandsHybrideAdaptive) getCyclePeriod() float64 {
	t := len(f.closes) - 1

	// Extend buffers to index t.
	for len(f.smoothBuf) <= t {
		f.smoothBuf = append(f.smoothBuf, 0.0)
	}

	for len(f.cycleBuf) <= t {
		f.cycleBuf = append(f.cycleBuf, 0.0)
	}

	for len(f.q1Buf) <= t {
		f.q1Buf = append(f.q1Buf, 0.0)
	}

	for len(f.i1Buf) <= t {
		f.i1Buf = append(f.i1Buf, 0.0)
	}

	for len(f.dpBuf) <= t {
		f.dpBuf = append(f.dpBuf, 0.0)
	}

	for len(f.instPeriodBuf) <= t {
		f.instPeriodBuf = append(f.instPeriodBuf, 6.0)
	}

	if t < 6 {
		return math.NaN()
	}

	prices := f.closes

	// 4-bar weighted smoother.
	f.smoothBuf[t] = (prices[t] + 2.0*prices[t-1] + 2.0*prices[t-2] + prices[t-3]) / 6.0

	// High-pass filter.
	alphaHP := f.alphaHP
	hpCoeff := (1.0 - 0.5*alphaHP) * (1.0 - 0.5*alphaHP)
	oneMinusAlpha := 1.0 - alphaHP

	f.cycleBuf[t] = hpCoeff*(f.smoothBuf[t]-2.0*f.smoothBuf[t-1]+f.smoothBuf[t-2]) +
		2.0*oneMinusAlpha*f.cycleBuf[t-1] - oneMinusAlpha*oneMinusAlpha*f.cycleBuf[t-2]

	// Quadrature component.
	f.q1Buf[t] = (0.0962*f.cycleBuf[t] + 0.5769*f.cycleBuf[t-2] -
		0.5769*f.cycleBuf[t-4] - 0.0962*f.cycleBuf[t-6]) *
		(0.5 + 0.08*f.instPeriodBuf[t-1])

	// In-phase component.
	f.i1Buf[t] = f.cycleBuf[t-3]

	// Smooth I and Q with EMA.
	if t > 6 {
		f.i1Buf[t] = 0.15*f.i1Buf[t] + 0.85*f.i1Buf[t-1]
		f.q1Buf[t] = 0.15*f.q1Buf[t] + 0.85*f.q1Buf[t-1]
	}

	// Compute delta phase.
	var dp float64

	if math.Abs(f.i1Buf[t]) > 1e-10 {
		dp = math.Atan(f.q1Buf[t] / f.i1Buf[t])
	} else {
		dp = f.dpBuf[t-1]
	}

	// Clamp delta phase.
	if dp < 0.1 {
		dp = 0.1
	}

	if dp > 1.1 {
		dp = 1.1
	}

	f.dpBuf[t] = dp

	// Median delta phase over 5 bars.
	var medianDP float64

	if t >= 10 {
		w := [5]float64{f.dpBuf[t-4], f.dpBuf[t-3], f.dpBuf[t-2], f.dpBuf[t-1], f.dpBuf[t]}
		// Sort 5 elements to find median.
		for i := 0; i < 4; i++ {
			for j := i + 1; j < 5; j++ {
				if w[j] < w[i] {
					w[i], w[j] = w[j], w[i]
				}
			}
		}

		medianDP = w[2]
	} else {
		medianDP = dp
	}

	// Instantaneous period.
	var dc float64

	if math.Abs(medianDP) > 1e-10 {
		dc = 6.2832/medianDP + 0.5
	} else {
		dc = f.instPeriodBuf[t-1]
	}

	// Clamp and smooth.
	if dc < 6.0 {
		dc = 6.0
	}

	if dc > 50.0 {
		dc = 50.0
	}

	f.instPeriodBuf[t] = 0.33*dc + 0.67*f.instPeriodBuf[t-1]

	return f.instPeriodBuf[t]
}

// Update updates the value of the indicator given the next sample.
// Returns the FRASMA2 value. Use UpdateAll() for all three outputs.
func (f *FractalBandsHybrideAdaptive) Update(sample float64) float64 {
	if math.IsNaN(sample) {
		return sample
	}

	f.mu.Lock()
	defer f.mu.Unlock()

	period := f.period
	windowSize := f.windowSize

	// Accumulate close history.
	f.closes = append(f.closes, sample)

	// Update Ehlers CyclePeriod.
	cp := f.getCyclePeriod()

	// Fill the FGDI window (period+1 elements).
	if f.windowCount < windowSize {
		f.window[f.windowCount] = sample
		f.windowCount++

		if f.windowCount < windowSize {
			return math.NaN()
		}

		f.primed = true
	} else {
		for i := 0; i < windowSize-1; i++ {
			f.window[i] = f.window[i+1]
		}

		f.window[windowSize-1] = sample
	}

	// FGDI computation over period+1 points (period segments).
	priceMax := f.window[0]
	priceMin := f.window[0]

	for k := 1; k < windowSize; k++ {
		if f.window[k] > priceMax {
			priceMax = f.window[k]
		}

		if f.window[k] < priceMin {
			priceMin = f.window[k]
		}
	}

	priceRange := priceMax - priceMin

	var fgdi float64

	if priceRange < 1e-10 {
		fgdi = 1.0
	} else {
		length := 0.0

		for i := 1; i < windowSize; i++ {
			normCur := (f.window[i] - priceMin) / priceRange
			normPrev := (f.window[i-1] - priceMin) / priceRange
			diff := normCur - normPrev
			length += math.Sqrt(diff*diff + f.invPeriodSq)
		}

		fgdi = 1.0 + (math.Log(length)+f.ln2)/f.logDenom
	}

	// Hurst exponent.
	hurst := 2.0 - fgdi
	if hurst < 0.01 {
		hurst = 0.01
	}

	trailDim := 1.0 / hurst
	beta := trailDim / 2.0

	// Adaptive normal_speed from CyclePeriod.
	var ns float64

	if math.IsNaN(cp) || cp < 1.0 {
		ns = float64(f.normalSpeedFallback)
	} else {
		ns = cp * f.nyquist
	}

	speed := int(math.Round(ns * beta))
	if speed < 1 {
		speed = 1
	}

	// FRASMA2: SMA of close over 'speed' bars ending at current position.
	nCloses := len(f.closes)
	if speed > nCloses {
		f.frasma2 = math.NaN()
		f.upperBand = math.NaN()
		f.lowerBand = math.NaN()

		return math.NaN()
	}

	smaSum := 0.0
	for k := nCloses - speed; k < nCloses; k++ {
		smaSum += f.closes[k]
	}

	frasma2Val := smaSum / float64(speed)

	// Deviation over the last period closes.
	sqSum := 0.0
	devStart := nCloses - period
	if devStart < 0 {
		devStart = 0
	}

	for k := devStart; k < nCloses; k++ {
		res := f.closes[k] - frasma2Val
		sqSum += res * res
	}

	deviation := 2.0 * math.Sqrt(sqSum/float64(period))

	// Fractal bands.
	bandMult := deviation * math.Pow(f.alpha, hurst)
	upperBand := frasma2Val + bandMult
	lowerBand := frasma2Val - bandMult

	f.frasma2 = frasma2Val
	f.upperBand = upperBand
	f.lowerBand = lowerBand

	return frasma2Val
}

// UpdateAll updates the indicator and returns all three outputs: frasma2, upperBand, lowerBand.
func (f *FractalBandsHybrideAdaptive) UpdateAll(sample float64) (float64, float64, float64) {
	frasma2 := f.Update(sample)

	f.mu.RLock()
	defer f.mu.RUnlock()

	return frasma2, f.upperBand, f.lowerBand
}

// Frasma2Value returns the last computed FRASMA2 value.
func (f *FractalBandsHybrideAdaptive) Frasma2Value() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.frasma2
}

// UpperBandValue returns the last computed upper band value.
func (f *FractalBandsHybrideAdaptive) UpperBandValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.upperBand
}

// LowerBandValue returns the last computed lower band value.
func (f *FractalBandsHybrideAdaptive) LowerBandValue() float64 {
	f.mu.RLock()
	defer f.mu.RUnlock()

	return f.lowerBand
}

// UpdateScalar updates the indicator given the next scalar sample.
func (f *FractalBandsHybrideAdaptive) UpdateScalar(sample *entities.Scalar) core.Output {
	frasma2, upper, lower := f.UpdateAll(sample.Value)

	const outputCount = 4

	output := make([]any, outputCount)
	output[0] = entities.Scalar{Time: sample.Time, Value: frasma2}
	output[1] = entities.Scalar{Time: sample.Time, Value: upper}
	output[2] = entities.Scalar{Time: sample.Time, Value: lower}

	if math.IsNaN(lower) || math.IsNaN(upper) {
		output[3] = outputs.NewEmptyBand(sample.Time)
	} else {
		output[3] = outputs.NewBand(sample.Time, lower, upper)
	}

	return output
}

// UpdateBar updates the indicator given the next bar sample.
func (f *FractalBandsHybrideAdaptive) UpdateBar(sample *entities.Bar) core.Output {
	v := f.barFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateQuote updates the indicator given the next quote sample.
func (f *FractalBandsHybrideAdaptive) UpdateQuote(sample *entities.Quote) core.Output {
	v := f.quoteFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}

// UpdateTrade updates the indicator given the next trade sample.
func (f *FractalBandsHybrideAdaptive) UpdateTrade(sample *entities.Trade) core.Output {
	v := f.tradeFunc(sample)

	return f.UpdateScalar(&entities.Scalar{Time: sample.Time, Value: v})
}
